File: src/engine/api/_period_detect.py
"""PERIOD DETECTION SERVICE (Part B) — the one place that answers
"which period does THIS document belong to?" from the document's own
evidence.

WHY THIS EXISTS
---------------
`documents.period_end_hint` is a CONFIRMATION channel: it means "a human
confirmed that THIS document belongs to THIS month". `stage_persist`
ranks it first and deliberately overrides its own detection because of
that meaning. The frontend was filling it with the DROP TARGET's date —
a number read off the UI, never off the document — so the engine
correctly discarded correct detections. The 2026-08-30 production audit
found every mismatched row carrying `hint == stored`, including a 2025
Carniprod trial balance filed under 2017-12.

The fix is semantic, and this module is its foundation: a hint-free
detection service that both the engine and the upload UI call, so the
two can never disagree, and whose SIGNATURE makes UI state
unrepresentable.

INVARIANTS
----------
W1  UI STATE IS NOT AN INPUT. `detect_period` takes exactly two
    keyword-only arguments — `extracted` and `filename`. No positional
    call, no ``**kwargs``. The currently-open period, the drop target
    and the last-used month have no way in.

W2  RANKED RESOLUTION, hint-free:
      1. in_document      — a period end the document itself states
      2. closing_balance  — a date sitting next to closing-balance
                            vocabulary in the document's own text
      3. filename         — via the engine's own helper, reused
      4. none             — ABSENT

W3  ABSENT != ZERO. "Not detected" is a first-class answer
    (`proposed_period_end=None`, `confidence=0.0`, `signal_used="none"`)
    that forces an explicit human choice. It is never today, never the
    open period, never a guess.

W4  TODAY IS NEVER A DETECTION. The engine's filename helper falls back
    to `date.today()` when it finds nothing; a date equal to today is
    therefore indistinguishable from "I found nothing" and is refused by
    every tier. The deliberate cost is one false negative: a document
    genuinely dated today resolves as ABSENT and asks the human. That is
    the safe direction — the alternative is exactly the silent
    misfiling this module exists to end.

W5  NO REIMPLEMENTATION. Every date is produced by the engine's own
    `_detect_period_end_from_filename` and clamped by its own
    `_sane_period_end`. This module contributes RANKING and LEXICAL
    NORMALIZATION only: it rewrites text into the vocabulary the engine
    helper already understands ("2025-12" -> "dec 2025") and lets the
    helper decide what that means, so month-end convention and sanity
    bounds keep living in exactly one place.

Python 3.9 — no `match`, no `X | Y` unions.
"""

from __future__ import annotations

import re
from datetime import date
from typing import Any, Dict, List, Optional, Tuple

#: The complete, closed set of signals `signal_used` can carry.
SIGNALS = ("in_document", "closing_balance", "filename", "none")

#: Confidence per tier. Stable numbers — the UI renders them and the
#: persisted detection record stores them, so they are contract, not
#: taste. Ordered strictly by tier so a lower-ranked signal can never
#: out-score a higher-ranked one.
CONFIDENCE = {
    "in_document": 0.95,
    "closing_balance": 0.85,
    "filename": 0.60,
    "none": 0.0,
}

#: Keys read off `extracted` for the in-document tier, in order. Both
#: the engine's snake_case parse output and a camelCase preview from the
#: frontend are accepted so one contract serves both callers.
_IN_DOCUMENT_KEYS = ("period_end", "periodEnd", "period_label", "periodLabel")

#: Keys read off `extracted` for the closing-balance tier, in order.
#: `closing_balance_date` is a resolved date; the rest are free text
#: lifted from the document (title block, column header, preamble).
_CLOSING_DATE_KEYS = ("closing_balance_date", "closingBalanceDate")
_CLOSING_TEXT_KEYS = (
    "closing_balance_label", "closingBalanceLabel",
    "header_text", "headerText",
    "period_text", "periodText",
    "document_text", "documentText",
    "preamble",
)

#: Phrases that mark a date as the period's CLOSING date rather than a
#: print timestamp or an invoice date. A date is only read from free
#: text when it sits inside the window opened by one of these.
_CLOSING_VOCABULARY = (
    # Romanian (with and without diacritics — exports vary)
    "balanta de verificare", "balanță de verificare",
    "la data de", "la datele de",
    "sold final", "solduri finale", "soldurile finale",
    "incheiat la", "încheiat la",
    "la sfarsitul", "la sfârșitul",
    "perioada", "luna",
    # International
    "closing balance", "balance as at", "balance as of",
    "as at", "as of", "period ended", "period end",
    "year ended", "year end",
)

#: Width of the text window opened after a vocabulary hit. Wide enough
#: to reach the date on the same line, narrow enough that an unrelated
#: date further down the page can't be captured.
_WINDOW = 80

#: Month number -> the tag the engine helper's own month table uses
#: (see `_detect_period_end_from_filename`). Used ONLY to rewrite
#: numeric year-month text into the helper's vocabulary; the helper
#: still decides which day of the month that means.
#: `test_month_tags_match_the_engine_helper` pins this against the
#: helper so it cannot drift.
_MONTH_TAGS = {
    1: "ian", 2: "feb", 3: "mar", 4: "apr", 5: "mai", 6: "iun",
    7: "iul", 8: "aug", 9: "sep", 10: "oct", 11: "noi", 12: "dec",
}


# ── engine helpers, imported lazily to avoid an import cycle ───────────
# `pipeline` imports this module at module level; importing back at
# module level would deadlock the import graph. The helpers are resolved
# on first use and cached.

_HELPERS = None  # type: Optional[Tuple[Any, Any]]


def _helpers():
    """(_detect_period_end_from_filename, _sane_period_end) — the
    engine's own. Never reimplemented, never forked."""
    global _HELPERS
    if _HELPERS is None:
        from .pipeline import (  # noqa: WPS433 — deliberate lazy import
            _detect_period_end_from_filename,
            _sane_period_end,
        )
        _HELPERS = (_detect_period_end_from_filename, _sane_period_end)
    return _HELPERS


# ── lexical normalization (W5) ────────────────────────────────────────


def _ym_to_tag(year: str, month: str) -> str:
    return "%s %s" % (_MONTH_TAGS[int(month)], year)


def _normalize_for_helper(text: str) -> str:
    """Rewrite date-ish text into the vocabulary the engine helper
    already parses. Purely lexical — it never decides a date.

    Runs ONLY as a second pass, after the helper has already been given
    the raw text and found nothing, so no currently-working filename can
    change meaning:

      "tb_2025-12.xlsx"  -> "tb dec 2025.xlsx"   (year-month)
      "TB 09-2024.xlsx"  -> "TB sep 2024.xlsx"   (month-year)
      "balanta_2025.xls" -> "balanta 2025.xls"   ('_' is a word char, so
                                                  the helper's \\b anchors
                                                  never fire on '_2025')
      "dec2025.xls"      -> "dec 2025.xls"       (letter/digit run split)
      "FY2025.xlsx"      -> "FY 2025.xlsx"
    """
    s = str(text)
    # YYYY-MM not followed by a day part.
    s = re.sub(
        r"(?<!\d)(20\d{2})[._\-](0?[1-9]|1[0-2])(?![._\-]?\d)",
        lambda m: _ym_to_tag(m.group(1), m.group(2)),
        s,
    )
    # MM-YYYY.
    s = re.sub(
        r"(?<!\d)(0?[1-9]|1[0-2])[._\-](20\d{2})(?!\d)",
        lambda m: _ym_to_tag(m.group(2), m.group(1)),
        s,
    )
    # '_' is a word character, so it suppresses the helper's \b anchors.
    s = s.replace("_", " ")
    # Split letter<->digit runs so "dec2025" / "FY2025" gain boundaries.
    s = re.sub(r"(?<=[A-Za-z])(?=\d)", " ", s)
    s = re.sub(r"(?<=\d)(?=[A-Za-z])", " ", s)
    return s


def _date_from_text(text: Optional[str]) -> Optional[str]:
    """Resolve an ISO period-end date out of arbitrary text, or None.

    Order: exact ISO date -> engine helper on the raw text -> engine
    helper on the lexically normalized text. Every candidate passes
    through `_sane_period_end`, and a candidate equal to today is
    refused (W4: today is the helper's "found nothing" value and can
    never be evidence)."""
    if text is None:
        return None
    s = str(text).strip()
    if not s or not any(ch.isdigit() for ch in s):
        # A date needs digits. Short-circuiting here also keeps the
        # helper's "no date pattern" warning out of the logs for the
        # ordinary case of prose with no date in it.
        return None

    detect_from_filename, sane = _helpers()
    today = date.today().isoformat()

    # 1. An exact ISO date (the common `parsed["period_end"]` shape).
    try:
        iso = date.fromisoformat(s[:10]).isoformat()
    except (TypeError, ValueError):
        iso = None
    if iso is not None:
        clamped = sane(iso)
        if clamped and clamped != today:
            return clamped
        if clamped == today:
            return None
        # Out of range (the legacy 2050-12-31 / 2115-03-31 rows): fall
        # through rather than propose a corrupt period.

    # 2/3. The engine's own text date extractor, raw then normalized.
    def _run(candidate_text: str) -> Optional[str]:
        value = sane(detect_from_filename(candidate_text))
        return value if value and value != today else None

    raw_value = _run(s)
    normalized = _normalize_for_helper(s)
    norm_value = _run(normalized) if normalized != s else None

    if raw_value and norm_value and norm_value != raw_value:
        # The normalized pass may only REFINE the month inside the year
        # the helper already found — never move the year.
        #
        # Refinement is the point: on "TB 09-2024.xlsx" the helper's
        # own patterns have no month-year rule, so it falls through to
        # its year-only rule and answers 2024-12-31 — filing a
        # September trial balance in December, which is precisely the
        # misfiling class this lane exists to end. Normalizing to
        # "TB sep 2024.xlsx" lets the HELPER read the month it already
        # knows how to read.
        #
        # The same-year guard is the safety rail: a lexical rewrite
        # must never be able to relabel a document's YEAR, so a
        # disagreeing year always loses to the raw pass.
        if norm_value[:4] == raw_value[:4]:
            return norm_value
        return raw_value
    return raw_value or norm_value


# ── tiers ─────────────────────────────────────────────────────────────


def _snippet(text: Any, limit: int = 140) -> str:
    """The literal evidence, whitespace-collapsed, for the UI's 'why'
    line. Never paraphrased — the reader must be able to find it in the
    document."""
    s = re.sub(r"\s+", " ", str(text)).strip()
    if len(s) > limit:
        s = s[: limit - 1].rstrip() + "…"
    return s


def _filename_candidate(filename: Optional[str]) -> Optional[Dict[str, Any]]:
    if not filename:
        return None
    value = _date_from_text(filename)
    if not value:
        return None
    return {
        "signal": "filename",
        "period_end": value,
        "evidence_snippet": _snippet(filename),
    }


def _in_document_candidate(
    extracted: Optional[Dict[str, Any]],
    filename_value: Optional[str],
) -> Optional[Dict[str, Any]]:
    """A period end the document itself states.

    A value identical to the filename's own answer is NOT emitted here.
    On the deterministic trial-balance path the engine seeds
    `parsed["period_end"]` with `_detect_period_end_from_filename(...)`
    (pipeline.py ~line 672), so an equal value carries zero independent
    evidence; reporting it as `in_document` would tell the operator the
    document said something it never said."""
    if not isinstance(extracted, dict):
        return None
    for key in _IN_DOCUMENT_KEYS:
        raw = extracted.get(key)
        if raw is None or raw == "":
            continue
        value = _date_from_text(raw)
        if not value or value == filename_value:
            continue
        return {
            "signal": "in_document",
            "period_end": value,
            "evidence_snippet": _snippet(raw),
        }
    return None


def _closing_balance_candidate(
    extracted: Optional[Dict[str, Any]],
) -> Optional[Dict[str, Any]]:
    """A date sitting next to closing-balance vocabulary in the
    document's own text ("BALANTA DE VERIFICARE la data de 31.12.2025",
    "Solduri finale 30.06.2025", "closing balance as at 2025-12-31").

    Free text is only mined INSIDE a vocabulary window: a bare date
    anywhere on the page (a print timestamp, an invoice date) is not
    evidence of the period and is deliberately ignored."""
    if not isinstance(extracted, dict):
        return None

    for key in _CLOSING_DATE_KEYS:
        raw = extracted.get(key)
        if raw:
            value = _date_from_text(raw)
            if value:
                return {
                    "signal": "closing_balance",
                    "period_end": value,
                    "evidence_snippet": _snippet(raw),
                }

    for key in _CLOSING_TEXT_KEYS:
        raw = extracted.get(key)
        if not raw:
            continue
        text = str(raw)
        lowered = text.lower()
        # Leftmost vocabulary hit first — reading order, and
        # deterministic regardless of the phrase list's own order.
        hits = []  # type: List[Tuple[int, str]]
        for phrase in _CLOSING_VOCABULARY:
            start = lowered.find(phrase)
            while start != -1:
                hits.append((start, phrase))
                start = lowered.find(phrase, start + 1)
        for start, _phrase in sorted(hits):
            window = text[start:start + _WINDOW]
            value = _date_from_text(window)
            if value:
                return {
                    "signal": "closing_balance",
                    "period_end": value,
                    "evidence_snippet": _snippet(window),
                }
    return None


# ── the service ───────────────────────────────────────────────────────


def detect_period(
    *,
    extracted: Optional[Dict[str, Any]],
    filename: Optional[str]
) -> Dict[str, Any]:
    """Propose the period this document belongs to, from its own
    evidence alone.

    W1: these two keyword-only parameters are the COMPLETE input. There
    is no channel — no argument, no ``**kwargs`` — through which the
    open period, the drop target or any other UI state could reach this
    decision. That absence is the fix.

    Args:
      extracted: the parse/preview of the document, or None. Read keys
        (all optional, snake_case or camelCase):
          period_end / period_label        -> in_document
          closing_balance_date             -> closing_balance
          closing_balance_label,
          header_text, period_text,
          document_text, preamble          -> closing_balance (mined
                                              inside a vocabulary window)
        Unknown keys are ignored, so callers may pass the whole parse.
      filename: the document's original filename, or None.

    Returns a dict with exactly these keys:
      proposed_period_end : ISO 'YYYY-MM-DD', or None when ABSENT
      confidence          : 0.0 .. 0.95, by tier (see CONFIDENCE)
      signal_used         : one of SIGNALS
      evidence_snippet    : the literal text that produced the answer,
                            for the UI's "why" line; None when ABSENT
      candidates          : every tier that resolved, in rank order —
                            [{signal, period_end, evidence_snippet}] —
                            so a caller can show the disagreement
                            ("the content says 2025-12, the filename
                            says 2017-12") instead of only the winner.
    """
    filename_candidate = _filename_candidate(filename)
    filename_value = filename_candidate["period_end"] if filename_candidate else None

    ranked = [
        _in_document_candidate(extracted, filename_value),
        _closing_balance_candidate(extracted),
        filename_candidate,
    ]
    candidates = [c for c in ranked if c]

    if not candidates:
        # W3 — ABSENT. Not today, not the open period, not a guess.
        return {
            "proposed_period_end": None,
            "confidence": CONFIDENCE["none"],
            "signal_used": "none",
            "evidence_snippet": None,
            "candidates": [],
        }

    winner = candidates[0]
    return {
        "proposed_period_end": winner["period_end"],
        "confidence": CONFIDENCE[winner["signal"]],
        "signal_used": winner["signal"],
        "evidence_snippet": winner["evidence_snippet"],
        "candidates": candidates,
    }

File: frontend/lib/periodDetect.ts
// periodDetect — the frontend client for the engine's hint-free period
// detection service (`/api/period/detect`, Part B), plus the entity
// identity read off the same header bytes.
//
// WHY THIS MODULE EXISTS
// ──────────────────────
// `documents.period_end_hint` is a CONFIRMATION channel. The engine's
// `stage_persist` ranks it ABOVE its own detection, deliberately, because
// the hint is supposed to mean "a human confirmed that THIS document
// belongs to THIS month". The workspace UI was filling it with the DROP
// TARGET's date — a number read off a row, never off the document — so
// the engine correctly discarded correct detections. The 2026-08-30
// production audit found every mismatched row carrying `hint == stored`.
//
// This module is what makes a REAL confirmation possible: it asks the
// document. The dialog shows the answer and its evidence; only what the
// human then confirms is allowed to become a hint.
//
// THE LAWS, restated for the client side
//   W1  UI STATE IS NOT AN INPUT. `buildDetectRequest` accepts a filename
//       and the document's own bytes. There is no parameter for the open
//       period, the drop target, or the last month used, and the engine's
//       POST model is `extra="forbid"`, so smuggling one is a 422 rather
//       than a silently-ignored field.
//   W3  ABSENT != ZERO. "Not detected" is a first-class answer
//       (`proposedPeriodEnd: null`, `signalUsed: "none"`). It is never
//       today, never the open period, never a guess — it forces the human
//       to choose.
//   W5  NO SECOND OPINION. The engine's service is the authority; this
//       module maps its answer verbatim. The in-browser reader below runs
//       ONLY when the engine is unreachable (it can be — see
//       lib/useBackendStatus.ts), and even then it reads the DOCUMENT,
//       never the UI, and says so via `origin: "browser"`.
//
// One deliberate difference from the engine, documented rather than
// hidden: the engine refuses any date equal to today because its filename
// helper returns today when it finds nothing, so the two are
// indistinguishable there. The browser readers below never fall back to
// today, so a genuine today-dated document survives this path. The
// direction of the difference is safe (more evidence, never less).

import { detectPeriodEndFromFilename, detectPeriodEndFromText } from "@/lib/detectPeriodEnd";
import { fetchPeriodFromApi } from "@/lib/activePeriod";
import { getSupabase } from "@/lib/supabase";

const API_URL =
  (import.meta.env.VITE_API_URL as string | undefined) ?? "http://127.0.0.1:8000";

/** The engine's closed set of signals (engine.api._period_detect.SIGNALS). */
export const PERIOD_SIGNALS = [
  "in_document",
  "closing_balance",
  "filename",
  "none",
] as const;

export type PeriodSignal = (typeof PERIOD_SIGNALS)[number];

export interface PeriodCandidate {
  signal: PeriodSignal;
  periodEnd: string;
  evidenceSnippet: string | null;
}

export interface PeriodDetection {
  /** ISO date, or null — which means ABSENT, not "no period". */
  proposedPeriodEnd: string | null;
  /** 0.95 / 0.85 / 0.60 / 0.0 by tier — the engine's numbers, unmodified. */
  confidence: number;
  signalUsed: PeriodSignal;
  /** The literal text the answer was read from — the "why" line. */
  evidenceSnippet: string | null;
  /** Every tier that resolved, rank order — lets the UI show a
   *  disagreement ("content says X, filename says Y"), not just a winner. */
  candidates: PeriodCandidate[];
  /** Who answered. "engine" = the shared service; "browser" = the engine
   *  was unreachable and the file was read here; "unavailable" = nobody
   *  could read anything (still ABSENT, never a guess). */
  origin: "engine" | "browser" | "unavailable";
}

/** The one shape of "not detected". Frozen: it is compared by value. */
export const ABSENT_DETECTION: PeriodDetection = Object.freeze({
  proposedPeriodEnd: null,
  confidence: 0,
  signalUsed: "none" as PeriodSignal,
  evidenceSnippet: null,
  candidates: [] as PeriodCandidate[],
  origin: "unavailable" as const,
});

export interface EntityIdentity {
  /** Company name as printed in the document, or null when absent. */
  name: string | null;
  /** Fiscal code, digits only ("RO 1234567" → "1234567"), or null. */
  cui: string | null;
  /** The literal line it was read from. */
  evidence: string | null;
}

export const ABSENT_ENTITY: EntityIdentity = Object.freeze({
  name: null,
  cui: null,
  evidence: null,
});

export interface DocumentInspection {
  detection: PeriodDetection;
  entity: EntityIdentity;
}

/** The POST body — W1's shape at the wire. */
export interface PeriodDetectRequest {
  filename: string | null;
  extracted: Record<string, string> | null;
}

// ─── the request ─────────────────────────────────────────────────────────

/**
 * Build the detect request. The only inputs are the document's name and
 * the document's own bytes; there is no third parameter and no rest
 * argument, so no caller can widen this into a channel for UI state.
 */
export function buildDetectRequest(input: {
  filename: string | null;
  headerText: string;
}): PeriodDetectRequest {
  const headerText = (input.headerText ?? "").trim();
  return {
    filename: input.filename ?? null,
    // ABSENT != ZERO: unreadable bytes send no `extracted` at all rather
    // than an empty object that would read as "I looked and found none".
    extracted: headerText ? { header_text: headerText.slice(0, 8000) } : null,
  };
}

function asSignal(raw: unknown): PeriodSignal {
  return (PERIOD_SIGNALS as readonly string[]).includes(String(raw))
    ? (raw as PeriodSignal)
    : "none";
}

function mapEngineAnswer(raw: Record<string, unknown>): PeriodDetection {
  const signal = asSignal(raw.signal_used);
  const proposed =
    typeof raw.proposed_period_end === "string" ? raw.proposed_period_end : null;
  const candidates: PeriodCandidate[] = Array.isArray(raw.candidates)
    ? (raw.candidates as Record<string, unknown>[])
        .filter((c) => c && typeof c.period_end === "string")
        .map((c) => ({
          signal: asSignal(c.signal),
          periodEnd: String(c.period_end),
          evidenceSnippet:
            typeof c.evidence_snippet === "string" ? c.evidence_snippet : null,
        }))
    : [];
  return {
    proposedPeriodEnd: signal === "none" ? null : proposed,
    confidence: typeof raw.confidence === "number" ? raw.confidence : 0,
    signalUsed: signal,
    evidenceSnippet:
      typeof raw.evidence_snippet === "string" ? raw.evidence_snippet : null,
    candidates,
    origin: "engine",
  };
}

/**
 * The in-browser reader. Runs ONLY when the engine cannot be reached, and
 * reads the same two things the engine reads — the document's text and
 * its filename — in the same rank order. Both helpers return null when
 * they recognise nothing, so this path cannot invent a date either.
 */
function detectInBrowser(filename: string | null, headerText: string): PeriodDetection {
  const fromText = detectPeriodEndFromText(headerText);
  if (fromText) {
    const line =
      headerText
        .split(/\r?\n/)
        .find((l) => /\d{4}/.test(l) && /balan|data|sold|perioad|period|as at|as of/i.test(l))
        ?.trim() ?? null;
    return {
      proposedPeriodEnd: fromText,
      confidence: 0.85,
      signalUsed: "closing_balance",
      evidenceSnippet: line ? line.slice(0, 160) : null,
      candidates: [
        { signal: "closing_balance", periodEnd: fromText, evidenceSnippet: line },
      ],
      origin: "browser",
    };
  }
  const fromName = detectPeriodEndFromFilename(filename);
  if (fromName) {
    return {
      proposedPeriodEnd: fromName,
      confidence: 0.6,
      signalUsed: "filename",
      evidenceSnippet: filename,
      candidates: [
        { signal: "filename", periodEnd: fromName, evidenceSnippet: filename },
      ],
      origin: "browser",
    };
  }
  return { ...ABSENT_DETECTION, candidates: [] };
}

/**
 * Ask the engine what month this document covers. Falls back to the
 * in-browser reader when the engine is unreachable, and to ABSENT when
 * neither can read anything.
 */
export async function detectPeriodFromEvidence(input: {
  filename: string | null;
  headerText: string;
}): Promise<PeriodDetection> {
  const body = buildDetectRequest(input);
  try {
    const sb = getSupabase();
    const token = sb
      ? (await sb.auth.getSession()).data.session?.access_token ?? null
      : null;
    if (!token) return detectInBrowser(input.filename, input.headerText);
    const res = await fetch(`${API_URL}/api/period/detect`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!res.ok) return detectInBrowser(input.filename, input.headerText);
    return mapEngineAnswer((await res.json()) as Record<string, unknown>);
  } catch {
    // Engine down / offline / CORS — the document still has evidence.
    return detectInBrowser(input.filename, input.headerText);
  }
}

// ─── reading the document's header ───────────────────────────────────────

/** `Blob.text` / `Blob.arrayBuffer` are recent; FileReader is not. Reading
 *  through the fallback keeps this working on older Safari — and in jsdom,
 *  whose File implements neither, which is where these tests run. */
function readFileAsText(file: File): Promise<string> {
  const blob = file as unknown as { text?: () => Promise<string> };
  if (typeof blob.text === "function") return blob.text();
  return new Promise((resolve, reject) => {
    const fr = new FileReader();
    fr.onload = () => resolve(typeof fr.result === "string" ? fr.result : "");
    fr.onerror = () => reject(fr.error ?? new Error("file read failed"));
    fr.readAsText(file);
  });
}

function readFileAsArrayBuffer(file: File): Promise<ArrayBuffer> {
  const blob = file as unknown as { arrayBuffer?: () => Promise<ArrayBuffer> };
  if (typeof blob.arrayBuffer === "function") return blob.arrayBuffer();
  return new Promise((resolve, reject) => {
    const fr = new FileReader();
    fr.onload = () =>
      resolve(fr.result instanceof ArrayBuffer ? fr.result : new ArrayBuffer(0));
    fr.onerror = () => reject(fr.error ?? new Error("file read failed"));
    fr.readAsArrayBuffer(file);
  });
}

/**
 * The document's title block — the first rows of a workbook / CSV, where
 * the company line and the reporting period live. Deliberately small:
 * the ledger rows below carry stray dates and no identity, and scanning
 * them would only add noise. Returns "" for formats the browser cannot
 * read (PDF, images) — the engine still gets the filename.
 */
export async function readDocumentHeader(file: File): Promise<string> {
  const lower = (file.name || "").toLowerCase();
  try {
    if (/\.(xlsx|xls)$/.test(lower)) {
      const XLSX = await import("xlsx");
      const wb = XLSX.read(await readFileAsArrayBuffer(file), { type: "array" });
      const parts: string[] = [];
      for (const sheet of wb.SheetNames.slice(0, 3)) {
        const rows = XLSX.utils.sheet_to_json<unknown[]>(wb.Sheets[sheet], {
          header: 1,
          blankrows: false,
        });
        for (const row of rows.slice(0, 40)) {
          parts.push((row as unknown[]).map((c) => (c == null ? "" : String(c))).join(" "));
        }
      }
      return parts.join("\n");
    }
    if (/\.(csv|tsv|txt)$/.test(lower)) {
      const text = await readFileAsText(file);
      return text.split(/\r?\n/).slice(0, 40).join("\n");
    }
  } catch {
    // Corrupt / unreadable — fall through to "" (the filename still counts).
  }
  return "";
}

// ─── entity identity ─────────────────────────────────────────────────────

/** Legal forms, as printed in Romanian exports (with and without dots). */
const LEGAL_FORM_TAIL =
  /\b(s\.?\s*r\.?\s*l|s\.?\s*a|s\.?\s*n\.?\s*c|s\.?\s*c\.?\s*s|p\.?\s*f\.?\s*a|kft|zrt|gmbh|ltd|llc|bv|nv|plc|ag|sp\.?\s*z\.?\s*o\.?\s*o)\b\.?\s*$/i;

/** Labels an export may put in front of the company name. */
const NAME_LABEL = /^(denumire|denumirea|societatea|firma|unitatea|company|entity)\s*[:\-]\s*/i;

/** "C.U.I. RO 1234567", "CIF: 1234567", "Cod fiscal 1234567". */
const CUI_RE = /\b(?:c\.?\s*u\.?\s*i\.?|c\.?\s*i\.?\s*f\.?|cod\s+fiscal|vat)\s*[:\-]?\s*(?:ro)?\s*(\d{2,10})\b/i;

/**
 * Read the company identity off a document's header text. Both fields are
 * independently optional: ABSENT != ZERO, so an unreadable header yields
 * nulls and the entity guard simply never fires — it never guesses.
 */
export function detectEntityFromHeader(text: string | null | undefined): EntityIdentity {
  if (!text) return { ...ABSENT_ENTITY };
  const lines = text.split(/\r?\n/).map((l) => l.replace(/\s+/g, " ").trim());

  let cui: string | null = null;
  for (const line of lines) {
    const m = line.match(CUI_RE);
    if (m) {
      cui = m[1];
      break;
    }
  }

  let name: string | null = null;
  let evidence: string | null = null;
  for (const raw of lines) {
    if (!raw || raw.length > 120) continue;
    const line = raw.replace(NAME_LABEL, "").trim();
    if (!line || !/[A-Za-zĂÂÎȘȚăâîșț]{2,}/.test(line)) continue;
    if (!LEGAL_FORM_TAIL.test(line)) continue;
    name = line;
    evidence = raw;
    break;
  }

  return { name, cui, evidence };
}

/** Digits only — "RO 1234567" and "1234567" are the same company. */
export function normalizeCui(raw: string | null | undefined): string | null {
  if (!raw) return null;
  const digits = String(raw).replace(/\D+/g, "");
  return digits.length >= 2 ? digits.replace(/^0+/, "") : null;
}

/**
 * A comparable company name: diacritics folded, legal form and the "S.C."
 * prefix dropped, punctuation removed. "S.C. Scandia RealEstate S.R.L."
 * and "SCANDIA REALESTATE SRL" normalize to the same string.
 */
export function normalizeEntityName(raw: string | null | undefined): string | null {
  if (!raw) return null;
  let s = String(raw)
    .normalize("NFD")
    .replace(/[̀-ͯ]/g, "")
    .toLowerCase();
  s = s.replace(NAME_LABEL, "");
  s = s.replace(/^s\.?\s*c\.?\s+/i, "");
  s = s.replace(
    /\b(s\.?\s*r\.?\s*l|s\.?\s*a|s\.?\s*n\.?\s*c|s\.?\s*c\.?\s*s|p\.?\s*f\.?\s*a|kft|zrt|gmbh|ltd|llc|bv|nv|plc|ag)\b\.?/g,
    " ",
  );
  s = s.replace(/[^a-z0-9]+/g, " ").trim();
  return s || null;
}

/**
 * Do these two identities describe different companies? Answers `false`
 * whenever it cannot tell — a guard that fires on absence would train the
 * user to click through it.
 */
export function entitiesConflict(
  a: EntityIdentity | null | undefined,
  b: EntityIdentity | null | undefined,
): boolean {
  if (!a || !b) return false;
  const ca = normalizeCui(a.cui);
  const cb = normalizeCui(b.cui);
  // A fiscal code is the identity; a renamed company keeps its CUI.
  if (ca && cb) return ca !== cb;
  const na = normalizeEntityName(a.name);
  const nb = normalizeEntityName(b.name);
  if (!na || !nb) return false;
  if (na === nb) return false;
  // One name containing the other is a shortening ("scandia food" vs
  // "scandia food holding"), not a different company.
  if (na.includes(nb) || nb.includes(na)) return false;
  return true;
}

/** Positive identities only. A period's company does not change, so a
 *  resolved name is worth keeping across month toggles; a NULL is not
 *  cached, because "no analysis yet" becomes "analysed" the moment the
 *  upload we are about to make finishes, and a cached null would silence
 *  the entity guard for the rest of the session. */
const _entityCache = new Map<string, EntityIdentity>();

/**
 * The identity of what a period ALREADY holds, read off that period's own
 * analysis (the engine's report carries the company it extracted). Null
 * when the period has no analysis yet — ABSENT, so the guard stays quiet.
 */
export async function resolvePeriodEntity(
  periodId: string | null | undefined,
): Promise<EntityIdentity | null> {
  if (!periodId) return null;
  const cached = _entityCache.get(periodId);
  if (cached) return cached;
  try {
    const res = await fetchPeriodFromApi(periodId);
    if (res.kind !== "ok") return null;
    // `res.data.statements` is already typed `Statements`; the previous
    // `as { statements?: Record<string, unknown> }` asserted a looser shape
    // over it, which is how `cui` — a field the blob does not contain —
    // read as a live lookup instead of a permanent null. See the `cui`
    // note on the `Statements` interface.
    const statements = res.data.statements;
    const name =
      statements && typeof statements.companyName === "string"
        ? statements.companyName
        : null;
    const cui =
      statements && typeof statements.cui === "string" ? statements.cui : null;
    if (!name && !cui) return null;
    const entity: EntityIdentity = { name, cui, evidence: null };
    _entityCache.set(periodId, entity);
    return entity;
  } catch {
    return null;
  }
}

/**
 * One read of the file, both guards' evidence. The dialog calls exactly
 * this, with exactly one argument — the file. That signature is the UI's
 * half of W1.
 */
export async function inspectDocument(file: File): Promise<DocumentInspection> {
  const headerText = await readDocumentHeader(file);
  const detection = await detectPeriodFromEvidence({
    filename: file.name ?? null,
    headerText,
  });
  return { detection, entity: detectEntityFromHeader(headerText) };
}
